### PR TITLE
[ doc ] cookbook for parsing

### PR DESCRIPTION
**Pull request by thomasdziedzic**
As part of the Idris Developer Meeting, I decided to scratch an itch by documenting the Text.Lexer/Text.Parser modules in a cookbook since I've had to figure out parsing using these modules around 3 times now.

I decided to include the original code for the lambda calculus parser and the calculator example in the docs so that folks could just include them rather than copying from the docs. If that isn't desirable, I can remove them.

### CODE REVIEW

File: docs/source/cookbook/Calculator.idr [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  term = do
+    num <- match CTNum
+    pure num
+
+  expr : Grammar state CalculatorToken True Double
+  expr = buildExpressionParser [
+    [ Infix (match CTMultiply >> pure (*)) AssocLeft
+    , Infix (match CTDivide >> pure (/)) AssocLeft
+    ],
+    [ Infix (match CTPlus >> pure (+)) AssocLeft
```
Comment (gallais) [inline docs/source/cookbook/Calculator.idr:91]: It's a stylistic choice but I find the following nicer:
```suggestion
    [ Infix ((+) <$ match CTPlus) AssocLeft
```
Thoughts?

Comment (gallais) [inline docs/source/cookbook/Calculator.idr:91]: It's a stylistic choice but I find the following nicer:
```suggestion
    [ Infix ((+) <$ match CTPlus) AssocLeft
```
Thoughts?

File: docs/source/cookbook/parsing.rst [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+The ``Text.Lexer`` module also reexports ``Text.Lexer.Core``, ``Text.Quantity`` and ``Text.Token``.
+
+``Text.Lexer.Core`` provides the building blocks of the lexer, including a type called
+``Recognise`` which is the underlying data type for the lexer. The other important function that this
+module provides is a ``lex`` which takes in a lexer and returns the tokens.
+
+``Text.Quantity`` provides a data type ``Quantity`` which can be used with certain lexers to specify
+how many times something is expected to appear.
+
+``Text.Token`` provides a data type ``Token`` that represents a parsed token, it's kind and the text.
```
Comment (gallais) [inline docs/source/cookbook/parsing.rst:39]:
```suggestion
``Text.Token`` provides a data type ``Token`` that represents a parsed token, its kind and the text.
```

Comment (gallais) [inline docs/source/cookbook/parsing.rst:39]:
```suggestion
``Text.Token`` provides a data type ``Token`` that represents a parsed token, its kind and the text.
```

File: docs/source/cookbook/parsing.rst [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+    %default total
+
+    data Expr = App Expr Expr | Abs String Expr | Var String | Let String Expr Expr
+
+    Show Expr where
+      show (App e1 e2) = "(" ++ show e1 ++ " " ++ show e2 ++ ")"
+      show (Abs v e) = "(\\" ++ v ++ "." ++ show e ++ ")"
+      show (Var v) = v
+      show (Let v e1 e2) = "let " ++ v ++ " = " ++ show e1 ++ " in " ++ show e2
```
Comment (gallais) [inline docs/source/cookbook/parsing.rst:79]: This should hopefully produce a nicer ouptut.
It's debatable whether we should bother given that the cookbook is about something else but the
output we get in the example program is a bit annoyingly overparenthesised so...
```suggestion
      showPrec d (App e1 e2) = showParens (d == App) (showPrec (User 0) e1 ++ " " ++ showPrec App e2)
      showPrec d (Abs v e) = showParens (d > Open) ("\\" ++ v ++ "." ++ show e)
      showPrec d (Var v) = v
      showPrec d (Let v e1 e2) = showParens (d > Open) ("let " ++ v ++ " = " ++ show e1 ++ " in " ++ show e2)
```

Comment (gallais) [inline docs/source/cookbook/parsing.rst:79]: This should hopefully produce a nicer ouptut.
It's debatable whether we should bother given that the cookbook is about something else but the
output we get in the example program is a bit annoyingly overparenthesised so...
```suggestion
      showPrec d (App e1 e2) = showParens (d == App) (showPrec (User 0) e1 ++ " " ++ showPrec App e2)
      showPrec d (Abs v e) = showParens (d > Open) ("\\" ++ v ++ "." ++ show e)
      showPrec d (Var v) = v
      showPrec d (Let v e1 e2) = showParens (d > Open) ("let " ++ v ++ " = " ++ show e1 ++ " in " ++ show e2)
```

**Review by gallais — COMMENTED** [review]
Github has somehow duplicated my comments?!

File: docs/source/cookbook/parsing.rst (hunk 2 above)
Comment (thomasdziedzic) [reply docs/source/cookbook/parsing.rst]: english is weird, thanks

File: docs/source/cookbook/parsing.rst (hunk 3 above)
Comment (thomasdziedzic) [reply docs/source/cookbook/parsing.rst]: I actually had pretty printed code in my original interpreter that I pulled this code from, but I decided to simplify it in here. I'll add your suggestion though. I also learned about showPrec and showParens which are built into Idris 2!

File: docs/source/cookbook/Calculator.idr (hunk 1 above)
Comment (thomasdziedzic) [reply docs/source/cookbook/Calculator.idr]: changed!

**Pull request merged**